pmlc/rt/device.h

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020, Intel Corporation
+
+#pragma once
+
+#include <memory>
+#include <stdexcept>
+#include <utility>
+
+#include "pmlc/rt/executable.h"
+
+namespace pmlc::rt {
+
+// Device represents a PlaidML device, capable of evaluating a PlaidML program.
+class Device {
+public:
+  // Returns the current Device, local to the current thread.
+  //
+  // If there is no current device, or the current device is of an unexpected
+  // type, an exception will be thrown.
+  //
+  // This is used to smuggle the current device through to runtimes that are
+  // directly invoked by lowered code, when that lowered code doesn't support
+  // the use of a passthrough parameter.
+  template <typename T>
+  static std::shared_ptr<T> current() {
+    std::shared_ptr<Device> dev = currentUntyped();
+    if (!dev) {
+      throw std::logic_error{"No current device established"};
+    }
+    std::shared_ptr<T> result = std::dynamic_pointer_cast<T>(std::move(dev));
+    if (!result) {
+      throw std::runtime_error{"Incompatible target/device combination"};
+    }
+    return result;
+  }
+
+  virtual ~Device() = default;
+
+  virtual std::unique_ptr<Executable>
+  compile(const std::shared_ptr<pmlc::compiler::Program> &program,
```

Comment:
It seems like only a `Program*` is needed here? Why would an `Executable` need to hold onto a `program`?

Reply:
This is just moving around & refactoring existing code... but I could go either way on this -- the counterargument is that since we're passing around Program with shared_ptr in other places, we should avoid passing it via a raw pointer, since it's entirely too easy to build a shared_ptr out of a raw pointer and get a double-delete.  (When do we switch to Rust, again?  :-)

Reply:
I guess what's needed here is a description of the intended lifetime and usage policy for these objects. My thinking was that the user would be in control over when a program, device, executable, or runtime would be destroyed. But perhaps it's better to use shared_ptr everywhere so that users don't have to be as careful (and we don't have to work out every single workflow). Overall it's cheap, so ya, let's stick with shared_ptr.